core/src/main/scala/com/avast/grpc/jsonbridge/ReflectionGrpcJsonBridge.scala

```diff
@@ -114,32 +99,50 @@ object ReflectionGrpcJsonBridge extends StrictLogging {
       method.invoke(null, inProcessChannel).asInstanceOf[AbstractStub[_]]
   }
 
-  private def createServiceHandlers[F[_]](ec: ExecutionContext)(inProcessChannel: ManagedChannel)(ssd: ServerServiceDefinition)(
-      implicit F: Async[F]): Map[String, HandlerFunc[F]] = {
+  private def createServiceHandlers[F[+ _]](ec: ExecutionContext)(inProcessChannel: ManagedChannel)(ssd: ServerServiceDefinition)(
+      implicit F: Async[F]): Map[GrpcMethodName, HandlerFunc[F]] = {
     val futureStubCtor = createFutureStubCtor(ssd.getServiceDescriptor, inProcessChannel)
     ssd.getMethods.asScala
       .filter(isSupportedMethod)
       .map(createHandler(ec)(futureStubCtor)(_))
       .toMap
   }
 
-  private def createHandler[F[_]](ec: ExecutionContext)(futureStubCtor: () => AbstractStub[_])(method: ServerMethodDefinition[_, _])(
-      implicit F: Async[F]): (String, HandlerFunc[F]) = {
-    val requestMessagePrototype = getRequestMessagePrototype(method)
-    val javaMethod = futureStubCtor().getClass
-      .getDeclaredMethod(getJavaMethodName(method), requestMessagePrototype.getClass)
-    val execute = executeRequest[F](ec)(futureStubCtor, javaMethod) _
-
-    val handler: HandlerFunc[F] = (json, headers) => {
+  private def handler[F[+ _]](requestMessagePrototype: Message, execute: (Message, Map[String, String]) => F[MessageOrBuilder])(
+      implicit F: Async[F]): HandlerFunc[F] = { (json, headers) =>
+    {
       parseRequest(json, requestMessagePrototype) match {
-        case Right(req) => execute(req, headers).map(resp => Right(printer.print(resp)))
-        case Left(status) => F.pure(Left(status))
+        case Right(req) =>
+          execute(req, headers)
+            .map(printer.print)
+            .map(Right(_): Either[BridgeError.Narrow, String])
+            .recover {
+              case e: StatusException =>
+                Left(BridgeError.RequestErrorGrpc(e.getStatus))
```

Comment:
IMHO it's not a good idea to drop the original exception here because it can contain interesting information that could be logged later. So I believe it should be passed to the `BridgeError`.

Reply:
as far as I understand the exception is just another representation of a gRPC `Status`, as if 1:1, but extends `Throwable`, to be usable by the `throw`/`catch` machinery on JVM

so not information lost here

Reply:
Ah, you're right - both exceptions don't allow to set an inner exception.